fixed unassign issues

SukhvirKooner opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
Fixes [#12837](https://github.com/learningequality/kolibri/issues/12837)
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->
All the reported issues have now been resolved:  

### ✅ **1. Unexpected Unassignments**  
- The action now retains assignments for linked PRs, regardless of variations in the link format.  
- The action also properly distinguishes between issues and pull requests, ensuring it doesn't act on PRs.  

### ✅ **2. Organization Member Unassignment**  
- The action now reliably differentiates external accounts from organization members, ensuring organization members are never unintentionally unassigned.  
<img width="1038" alt="Screenshot 2025-02-07 at 11 57 03 PM" src="https://github.com/user-attachments/assets/52d6fed8-ab63-411e-bd72-7401e5934d99" />


### ✅ **3. Slack Notification Format Improvement**  
- The Slack notification format has been optimized to minimize space usage and improve readability.  
- 
<img width="844" alt="image" src="https://github.com/user-attachments/assets/0fce168a-1ddf-4229-9214-2de0d71bad8f" />


Below are some Screenshots of the Testing i have done:
 


…

## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

…

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

…

MisRob commented:
Hi @SukhvirKooner, thank you for these updates. I have just finished testing another action and will move toward reviewing and re-testing this one.

MisRob commented on scripts/unassign-inactive-issues.js:
```diff
@@ -35,44 +35,213 @@ async function getAllIssues(github, owner, repo) {
+          (event.event === 'connected' && event?.source?.issue?.pull_request?.merged === false)
+        ) {
+          try {
+            let prNumber = event?.source?.issue?.number;
```
Only to understand , what was the motivation for reaching into `issue.number` rather than rely solely on `pull_request.number`?

SukhvirKooner replied:
The code uses `issue.number` instead of `pull_request.number` because in GitHub's event payload structure, every pull request is also an issue, but not every issue is a pull request. Some webhook events provide only the `issue` object, even for pull requests, without explicitly including a `pull_request` field. This ensures that the PR number is consistently retrievable across different event types.

**However, there is still one case where the PR number could not be fetched—when a PR is linked through the GitHub GUI.** Due to this limitation, the following branch was added to handle that scenario:
```

console.log('founded pr linked in the issue');
linkedPRs.add(1);

```
This addition fixed the issue by ensuring that linked PRs are properly tracked even when they don’t appear in the expected payload structure.

I explored every possible method to retrieve the PR number when it is linked through the GitHub GUI, but none of them worked reliably. This approach was the most effective solution I found.

Hope this clears both your queries!

MisRob replied:
I see - would you comment on these places in the code? I think this is one of those situations when it's hard to figure out from code, and a bit more explanation will be appreciated by future maintainers.

SukhvirKooner replied:
Will do! I'll comments as well.

MisRob commented on scripts/unassign-inactive-issues.js:
```diff
@@ -35,44 +35,213 @@ async function getAllIssues(github, owner, repo) {
+              
+              if (pr && pr.state === 'open') {
+                console.log(`Found valid linked PR #${prNumber} (${pr.state})`);
+                linkedPRs.add(prNumber); // Use add() instead of push()
```
```
// Use add() instead of push()
```

What is this comment trying to say?

SukhvirKooner replied:
Apologies for the oversight! These comments were just for my reference while working on the code. They helped me track changes and considerations during development, but I accidentally forgot to remove them before making the PR. I will clean them up.

MisRob replied:
No problem

MisRob commented on scripts/unassign-inactive-issues.js:
```diff
@@ -35,44 +35,213 @@ async function getAllIssues(github, owner, repo) {
+                console.log(`Found valid linked PR #${prNumber} (${pr.state})`);
+                linkedPRs.add(prNumber); // Use add() instead of push()
+              }
+            }else{
```
What is this branch trying to do?

SukhvirKooner replied:
there is one case where the PR number could not be fetched—when a PR is linked through the GitHub GUI. Due to this limitation, the following branch was added to handle that scenario:
```

console.log('founded pr linked in the issue');
linkedPRs.add(1);

```
This addition fixed the issue by ensuring that linked PRs are properly tracked even when they don’t appear in the expected payload structure.

I explored every possible method to retrieve the PR number when it is linked through the GitHub GUI, but none of them worked reliably. This approach was the most effective solution I found.

MisRob replied:
Ah okay, I was wondering if this is some sort of workaround. Thanks for following-up. Aligned with my other feedback, I'd appreciate if you could add an inline code about this too.

SukhvirKooner replied:
Sure thing! I'll include the inline code for this as well.

MisRob commented on scripts/unassign-inactive-issues.js:
```diff
@@ -35,44 +35,213 @@ async function getAllIssues(github, owner, repo) {
     }
 
-    console.log('Slack webhook URL exists:', !!slackWebhookUrl);
+    // Method 2: Search for PRs that mention this issue
```
Using a query seems to be a nice addition :+1: However I think we'd need to combine it with regex (as in Method 3) that would filter these PRs to only those that close/fix/etc. the issue, right?

SukhvirKooner replied:
Absolutely! Combining GitHub’s search with a regex filter ensures that you catch only those PRs that explicitly include "closes #123," "fixes #123," or similar phrases. Since GitHub Search doesn’t support advanced regex in its query, the best approach is to:

1. Search for open PRs that mention the issue number in the title or body.
2. Apply a local regex check (like the one we already use in Method 3) to verify that the PR text truly includes a closing keyword alongside the issue number.

This effectively filters out PRs that merely mention #123 without actually linking or closing it.

**So, should I add the regex check into Method 2, given that the logic already exists in Method 3?**

MisRob replied:
Yes that's what I had in mind. I think the regex was a really nice idea that will make this action precise, so let's use it in Method 2. For method 3, see the other comment.

SukhvirKooner replied:
Got it! I'll go ahead and implement that for Method 2.

MisRob commented on scripts/unassign-inactive-issues.js:
```diff
@@ -35,44 +35,213 @@ async function getAllIssues(github, owner, repo) {
+      console.log('Search API error:', searchError.message);
+    }
+
+    // Method 3: Check issue body for PR references
```
I quite don't understand the need for this last method - typically it is in a PR description where authors write "closes [issue no.]", rather than in a related issue, right? What situation did you mean to represent with this method?

(by the way, I think it's valuable to use this regex, however if I can understand, it seems it'd be more appropriate to use it as part of Method 2?)

SukhvirKooner replied:
Method 3 is really just a “catch-all” for any time someone might have manually linked the PR from the issue side. It’s definitely less common than the “closes #123” usage in a PR body, but it can still happen. For example, a contributor might comment or edit the issue description to say “See PR #123,” or “We’re working on this in PR #123,” which isn’t necessarily in the PR’s description but rather in the issue’s own text.

In most projects, people indeed put “closes #IssueNumber” in the PR description, so Method 3 won’t fire often. But it’s there to handle that edge case where the issue body itself references the PR directly. If your team never links PRs from the issue side, I could omit Method 3 or move the regex logic to searching the PR descriptions (Method 2) as you suggested. It’s mainly about being extra thorough.

SukhvirKooner replied:
No Problem! I'll do that.

MisRob commented on scripts/unassign-inactive-issues.js:
```diff
@@ -96,198 +265,128 @@ module.exports = async ({github, context, core}) => {
-      }
-    };
-
     // Get all issues using pagination
```
Seems to be an obsolete comment

MisRob commented:
Hi @SukhvirKooner, thank you for these updates. It's shaping up nicely. I finished code review and left a few clarification questions and possibly requests. Let me know your thoughts. After we wrap up code review, I will merge and re-test.

SukhvirKooner commented on scripts/unassign-inactive-issues.js:
```diff
@@ -35,44 +35,213 @@ async function getAllIssues(github, owner, repo) {
+
+      
+      for (const event of timelineEvents) {
+        if (
```
Sure!

MisRob commented:
Hi @SukhvirKooner thanks a lot for all the clarifications. I will get back to this next week to read everything in detail and we can then decide on next steps.

MisRob commented:
Hi @SukhvirKooner, I left few last notes. After addressing them, would you be able to test as many scenarios as possible one more time in your environment? And if it works, we can merge, and I will give it another try in our repos. Thanks a lot for patient work in the area that has many nuances to address, seems we're getting close.

MisRob commented:
@SukhvirKooner oh and it seems there's a conflict, could you merge `main` and resolve it?

SukhvirKooner commented:
> Hi @SukhvirKooner, I left few last notes. After addressing them, would you be able to test as many scenarios as possible one more time in your environment? And if it works, we can merge, and I will give it another try in our repos. Thanks a lot for patient work in the area that has many nuances to address, seems we're getting close.

Sounds good! I'll address the notes and run a thorough round of testing. Thanks for the feedback!

SukhvirKooner commented:
Hi @MisRob,
I’ve added the regex in Method 2 and tested it. Also, I’ve removed Method 3 along with some unnecessary comments. I’m almost done testing all the scenarios.

If you have any additional scenarios or edge cases that you think should be covered, please let me know.

MisRob commented:
Thanks @SukhvirKooner, nothing comes to my mind, except all the situations we've talked about. Let me know when you think this is ready.

SukhvirKooner commented:
Hi @MisRob ,
I've tested all the scenarios we discussed. Everything looks good on my end, and you can go ahead and merge this.

MisRob commented:
Hi @SukhvirKooner, I did first re-testing and we still have some trouble with recognizing core team members. On these two issues, both MisRob and MisRob2 got unassigned, whereas only MisRob2 should be unassigned. MisRob is core team account.

![Screenshot from 2025-03-18 19-43-08](https://github.com/user-attachments/assets/7138026c-63bc-44c4-919e-702b0515a0fc)

Would taking inspiration[ from this action](https://github.com/learningequality/.github/blob/main/scripts/community-contribution-label.py#L5-L20) be helpful? I tested it some time ago and it worked well in this regard. It uses this endpoint https://docs.github.com/en/rest/collaborators/collaborators

SukhvirKooner commented:
Hi @MisRob,

Thanks for the detailed update! I had tested it earlier and just did a retest on my end—it’s working fine for me. I’ve attached screenshots below for reference.


<img width="510" alt="image" src="https://github.com/user-attachments/assets/91785e99-d554-4d00-a8fd-eab97dded6cd" />

- [Testing issue](https://github.com/Testing0179/test-actions/issues/21)

That said, I’ll definitely look into this further to see if there’s anything we might be missing in your case.

MisRob commented:
Hi @SukhvirKooner, yes I believe you tested it all, and thanks for double-checking now. I just think there's differences between your test environment and ours that we will need to fine-tune. Few things we can chat about:

- Do you know what's the difference between the methods you use and the collaborators API?
- Is there some additional logging we can add so I can provide you more information?
- Could there be some token permissions missing on our end?

MisRob commented:
@SukhvirKooner I found the log of one of those troublesome unassignments


![Screenshot from 2025-03-19 07-32-34](https://github.com/user-attachments/assets/d8720bef-245f-4d96-8690-e5dec2f0d298)

Notice the problematic `"MisRob is not an organization member"`.

Also I am seeing a deprecation notice that would be good to deal with if we can, but that's lesser priority.

MisRob commented:
@SukhvirKooner let me know any ideas about how I can assist you here.

Other than that I haven't noticed any problems, and Slack works too! So hopefully this is a last hiccup.

SukhvirKooner commented:
1. Difference Between Methods:
   - Current Method:
     Uses the organization membership check (and repository ownership) to determine if a user is "internal".
     This method may miss external collaborators who have been added as collaborators but are not members of the organization.
     
   - Collaborators API Method:
     Directly queries the /repos/{owner}/{repo}/collaborators/{username} endpoint.
     A 204 status indicates the user is a collaborator ("internal"), while a 404 indicates they are not (i.e., "external").
     
   ==> As a member of the core team (internal collaborator), you should not have been unassigned.

2. Additional Logging Suggestions (to add in your code):
   To help diagnose the issue, consider adding logs like these:
   
  ```
const checkUserMembership = async (owner, repo, username, github) => {
  if (!username) {
    console.error("Invalid username provided to checkUserMembership");
    return false;
  }
  
  try {
    console.log(`Fetching repository details for ${owner}/${repo} to check if ${username} is the owner...`);
    const repoDetails = await github.rest.repos.get({
      owner,
      repo
    });
    console.log(`Repository owner is ${repoDetails.data.owner.login}; comparing with ${username}`);
    
    if (repoDetails.data.owner.login === username) {
      console.log(`${username} is the repository owner.`);
      return true;
    }
    
    console.log(`Checking if ${username} is a member of organization ${owner}...`);
    try {
      const membershipResponse = await github.rest.orgs.getMembershipForUser({
        org: owner,
        username: username
      });
      console.log(`${username} is confirmed as an organization member. Membership response:`, membershipResponse.data);
      return true;
    } catch (orgError) {
      console.error(`${username} is not an organization member. Org membership error:`, orgError.response?.status, orgError.response?.data);
      return false;
    }
  } catch (error) {
    console.error(`Error checking user membership for ${username}:`, error.response?.status, error.response?.data);
    return false;
  }
};
```
In this updated version, I’ve added logging for:

- The repository details retrieval:
- To verify the owner and check that the correct username is being compared.

The organization membership check:
To log the membership response if the check is successful, and to log detailed error information if the check fails.
   
3. Necessary Token Permissions:
   - "repo" scope: Needed for accessing repository details (including collaborators) and updating issues.
   - "read:org" scope: Required for checking organization membership using endpoints like orgs.getMembershipForUser.
   
Ensure your token in all environments has these permissions to allow the API calls to work correctly.

SukhvirKooner commented:
@MisRob Please let me know if you'd like to implement the membership check method that was suggested. I can make those changes and raise a PR. **Additionally, I'll look into the deprecation issue.**

MisRob commented:
Thanks @SukhvirKooner, this is great. No you don't need to change that implementation yet, but feel free to add this logging, or I can do it some time later. Then we will know more and can decide what to do next.

I think those permissions should be fine but let me double check meanwhile.

MisRob commented:
Okay @SukhvirKooner so it really was the permission. It works now as expected https://github.com/learningequality/test-actions/issues/49#issuecomment-2737033866. Apologies for the hassle, it caught me by surprise because the other action for detecting membership doesn't need so many permissions like this one. So no changes needed in this area. Let me know how things look like with the deprecation issue and I can check then one last time.

SukhvirKooner commented:
Hey @MisRob,

I've fixed the deprecation issue. Should I open a new issue first and then submit a PR?

SukhvirKooner commented:
Hi @MisRob ,
I’ve opened a PR #20  for this! Looking forward to getting it merged.
